addons/mail/static/src/core/common/model/make_store.js

```diff
@@ -0,0 +1,314 @@
+/* @odoo-module */
+
+import { reactive } from "@odoo/owl";
+import {
+    IS_FIELD_DEFINITION_SYM,
+    IS_RECORD_FIELD_SYM,
+    IS_RECORD_SYM,
+    STORE_SYM,
+    modelRegistry,
+    _0,
+    isField,
+    isMany,
+    isRelation,
+} from "./misc";
+import { Store } from "./store";
+import { onChange } from "@mail/utils/common/misc";
+import { RecordList } from "./record_list";
+import { Record } from "./record";
+
+export function makeStore(env) {
+    const recordByLocalId = reactive(new Map());
+    const dummyStore = new Store();
+    dummyStore.recordByLocalId = recordByLocalId;
+    dummyStore.env = env;
+    const _ = { store: dummyStore };
+    Record.store0 = dummyStore;
+    const Models = {};
+    for (const [name, _OgClass] of modelRegistry.getEntries()) {
+        /** @type {typeof Record} */
+        const OgClass = _OgClass;
+        if (_.store[name]) {
+            throw new Error(`There must be no duplicated Model Names (duplicate found: ${name})`);
+        }
+        // classes cannot be made reactive because they are functions and they are not supported.
+        // work-around: make an object whose prototype is the class, so that static props become
+        // instance props.
+        /** @type {typeof Record} */
+        const Model = Object.create(OgClass);
+        // Produce another class with changed prototype, so that there are automatic get/set on relational fields
+        const Class = {
+            [OgClass.name]: class extends OgClass {
+                [IS_RECORD_SYM] = true;
+                constructor() {
+                    super();
+                    const this0 = this;
+                    this0._proxyUsed = new Set();
+                    this0._updateFields = new Set();
+                    this0._0 = this0;
```

Comment:
The numbering system doesn't really help for me, combined with `this` keyword it makes reading a line impossible without checking 3 different places for the meaning of a variable.

Reply:
"proxy", "internalProxy" and "fullProxy" were not much better, because it requires understanding the whole code to see which proxy they refer to and why it's important to use one or the other.

For pure correctness of the code, we don't need to understand which one we need. So it's not great they make the code very verbose for performance reasons. Using a numbering system allows to reduce the noise, while also showing more clearly each flavours is on a different level.

Of course, we could document better the different levels.